### DISCUSSION TITLE
Filesystem Cache

### DISCUSSION THREAD

**Discussion by wan9chi** [rolldown/rolldown#802] [General]
It's a much-requested feature in rollup: https://github.com/rollup/rollup/issues/2182. Although rollup only supports in-memory cache (`--watch`).

For rolldown, I think filesystem cache is more important than in-memory cache since it's meant to be a production bundler for vite.

I have some primitive ideas to try but I want to ask if this is interested before digging in.

**Comment by hyfdev** [comment]
Thanks for caring about it. Filesystem Cache is definitely something we are going to think about. However, it's a little bit early to say how we gonna do it. I mean we have built abstraction of file system, which might be a good start. But we don't have a cache design both for memory and file system so far.

**Comment by wan9chi** [comment]
Hi @hyf0,

In the implementation of `--watch`, rollup starts from scratch everytime but caches and reuses results of individual modules: https://github.com/rollup/rollup/issues/191#issuecomment-224986002.

I like this method, and think it can also be a good start for fs cache as it's easy to reason about and non-invasive to the existing workflow.

For rolldown, if we are taking this approach, I think it would mean caching and reusing

https://github.com/rolldown/rolldown/blob/a031226673d37deeebc3f73e3b0df15887b5486c/crates/rolldown/src/module_loader/normal_module_task.rs#L75

What do you think of this approach? Would you be interested in a PoC implementation?  I've already had some success on (de)serializing oxc ast nodes and semantic analysis results.

**Comment by wan9chi** [comment]
Deserializable oxc types: https://github.com/oxc-project/oxc/discussions/3079

**Reply by Boshen** [reply]
@branchseer Let me DM the Rolldown team on this.

I wonder whether persistent cache is useful given Rolldown is going to be so fast 🤔 People wanted it because Rollup is slow.

Oxc can parse (and eventually transform) files really fast, e.g. running the linter without any rules

```
pnpx oxlint@latest -A all
Finished in 5.3s on 70392 files with 0 rules using 12 threads.
```

This means we should persist other artifacts in Rolldown instead of the ASTs?

**Reply by wan9chi** [reply]
I just did a quick bench against `NormalModuleTask::scan` and scan output deserialization: https://github.com/branchseer/rolldown/blob/deser_bench/crates/rolldown/benches/scan.rs.

On my machine it took **26.447 ms** to scan a 1.3MB js module, and **8.4845 ms** to deserialize the scan output using `bincode2`. That's about **3** times as fast as a module scan.

I agree that its usefulness depends on how fast rolldown can become. Reducing 300ms to 100ms is not as useful as reducing 3s to 1s.

The bench is a very rough one and leaves some questions to be answered:
- `rkyv` is said to [be faster that bincode](https://rkyv.org/faq.html#isnt-this-very-unsafe-if-you-access-untrusted-data). I wonder how much faster we can get in this case.
- How much IO time can be saved? We can read all caches from a single db file, maybe memory-mapped, while source files scatter across the disk. Although cache invalidation still requires reading source file metadata from the fs catalog.

**Reply by Boshen** [reply]
Most js files are just a few hundred lines, so the difference may not be this large.

---

This discussion is about persistent cache for the whole build pipeline, so I recommend looking into other data to persist before going down the rabbit whole of persisting the AST, which requires a lot of effort and infrastructure.

**Reply by wan9chi** [reply]
I choose the scan part to start with because it's the first step and mostly pure, so caching it requires little change to existing code. Rolldown is still evolving rapidly after all.

But yeah I think I should profile the whole build process to find what parts are worth caching. Then again maybe it's too early to get a meaningful result.

**Comment by overlookmotel** [comment]
@hyf0 Just wanted to check how much of a gain you think a persistent cache for ASTs would be.

For the sake of argument, let's say the following is possible:

* Write out AST + semantic data for a JS/JSX/TS file to disk (either pre- or post-transform) with minimal serialization overhead.
* Read it back from disk into memory with *zero deserialization cost*. Just load into memory and cast a pointer.

*If* this were possible, would it be a large gain for Rolldown? Or is parsing/semantic/transform fast enough anyway that disk caching isn't particularly worthwhile?

I'm asking because I'm thinking about a custom allocator for Oxc (and accompanying custom `Box` and `Vec`). Design would be influenced by what use cases are most important to support. Some possible designs could support the above, though there are trade-offs, so am wondering if this is a priority use case or not.

**Reply by hyfdev** [reply]
> If this were possible, would it be a large gain for Rolldown?

It should takes advantages on the very large scale project, but it only can be sure with practices.

For normal caching, I guess we are just comparing the performance between oxc-parser and `serde_json/rkyv`? From what Boshen said, differences would be probably subtle. https://github.com/rolldown/rolldown/discussions/802#discussioncomment-9219661

In the bigger picture, the caching might doesn't need to touch ast and will only need to cache the codegen result of ast.

---

So I guess you don't need to take the caching ast as a important part while constructing a custom allocator.

**Reply by overlookmotel** [reply]
Thanks for coming back. Just to clarify:

What I have in mind is a *no-serialization implementation*. AST types would be stored in allocator in a format where you can just dump the allocator's memory buffer to disk with no serialization step - literally just copy memory to disk with no processing. In other direction, loading it back into memory is also purely load bytes from disk into memory, with no deserialization step. No overhead of serde/rkyv.

So the comparison isn't parser vs serde, but parser vs disk IO.

I know this is an unusual approach, but arena allocator (if built in a certain way) allows some stuff you can't usually do. So, in the context of this question, try to forget everything you know about how things usually work! Assume that ser/deser is 100% free.

I'm not familiar with Rolldown's internals, but don't you need to work with ASTs in order to link files together to build chunks? (rather than working with compiled JS code)

Does any of that affect the answer?

**Comment by e111077** [comment]
Chiming in here to suggest a cache model on top of a persistent cache.

Airbnb currently uses the Metro bundler but jerry rigged for web projects instead of React Native. Metro uses a [multi-layered cache](https://developerayo.github.io/metro/docs/en/caching) where you can provide multiple cache sources. We configure it to check for a memory cache, if miss then a persitent disk cache, then if that's a miss too, check a shared, persisted, read-only S3 bucket for a cache so that initial builds across the company can be somewhat warm.

_Update:_ So far we found that rolldown would beat Metro in cold, cache-less builds, but in everyday use with our caches existing in s3 and locally, Metro would greatly outperform rolldown end-to-end. And in interactive modes with rolldown in-memory cache and metro in-memory cache, their performance was _functionally_ equivalent. We find persistent caching an important requirement in our bundler largely for CI performance / cost and clean local dev builds

**Comment by shalldie** [comment]
Filesystem cache is very important to me, and I hope rolldown can implement it 😺

**Reply by IWANABETHATGUY** [reply]
Do you mean persistent cache? We have an issue to track it. We will implement that feature in the future, but since in most scenarios rolldown is fast enough, the priority may not be too high.

**Reply by shalldie** [reply]
Yes persistent cache. My scenario involves a monorepo with numerous projects. If there is persistent caching, it would be easier to convince the team to upgrade the build tool to Vite. Could you please provide the issue url? I haven't been able to find it.

**Reply by IWANABETHATGUY** [reply]
https://github.com/rolldown/rolldown/issues/4090
